### PR TITLE
51 - feed metadata into application header

### PR DESCRIPTION
**Pull request by caravinci**
Makes the application header come to life ⚡

### CODE REVIEW

File: components/pages/Applications/ApplicationForm/Header/Details.tsx [hunk 1]
```diff
@@ -26,7 +28,7 @@ const HeaderDetails = ({
         margin: 0 0 5px;
       `}
     >
-      <Link>My Applications</Link>: {appId.toUpperCase()}
+      <Link href={APPLICATIONS_PATH}>My Applications</Link>: {appId.toUpperCase()}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Header/Details.tsx:31]: making the breadcrumb link clickable, to go back to dashboard/management

File: components/pages/Applications/ApplicationForm/Header/index.tsx [hunk 2]
```diff
@@ -24,7 +33,12 @@ const ApplicationHeader = ({ appId = 'none' }): ReactElement => {
           width: 100%;
         `}
       >
-        <Details appId={appId} {...applicationDetails} />
+        <Details
+          appId={appId}
+          applicant={applicant}
+          createdAt={format(new Date(createdAtUtc), DATE_FORMAT)}
+          lastUpdated={format(new Date(lastUpdatedAtUtc), DATE_FORMAT + ' h:m aaaa')}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Header/index.tsx:40]: This is where we pass actual data now into the header.

File: components/pages/Applications/ApplicationForm/Header/index.tsx [hunk 3]
```diff
@@ -24,7 +33,12 @@ const ApplicationHeader = ({ appId = 'none' }): ReactElement => {
...
+        />
 
         <Progress />
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Header/index.tsx:43]: we'll pass the `application state` here

File: components/pages/Applications/ApplicationForm/index.tsx [hunk 4]
```diff
@@ -1,13 +1,34 @@
...
+    })
+      .then(({ data }: { data: Record<string, any> }) => setAppData(data))
+      .catch((error: Error) => {
+        // TODO dev logging, errors should not be shown to user
+        console.error(error);
+      });
+  }, []);
+
+  return isLoading || Object.values(appData).length < 1 ? (
+    <DnaLoader />
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/index.tsx:28]: add a temporary loader, to be improved upon at #49

File: components/pages/Applications/ApplicationForm/index.tsx [hunk 5]
```diff
@@ -1,13 +1,34 @@
...
 const ApplicationForm = ({ appId = 'none', isAdmin = false }): ReactElement => {
-  return (
+  const [appData, setAppData] = useState({});
+  const { fetchWithAuth, isLoading } = useAuthContext();
+
+  useEffect(() => {
+    fetchWithAuth({
+      url: `${API.APPLICATIONS}/${appId}`,
+    })
+      .then(({ data }: { data: Record<string, any> }) => setAppData(data))
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/index.tsx:20]: it puts the lotion in the basket 🐩

File: components/pages/Applications/ApplicationForm/index.tsx [hunk 6]
```diff
@@ -1,13 +1,34 @@
...
+        console.error(error);
+      });
+  }, []);
+
+  return isLoading || Object.values(appData).length < 1 ? (
+    <DnaLoader />
+  ) : (
     <>
-      <ApplicationHeader appId={appId} />
+      <ApplicationHeader data={appData} />
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/index.tsx:31]: passes the data to the header

File: components/pages/Applications/index.tsx [hunk 7]
```diff
@@ -1,14 +1,15 @@
...
 import { isDacoAdmin } from 'global/utils/egoTokenUtils';
 import { useAuthContext } from 'global/hooks';
 import DefaultPageLayout from 'components/DefaultPageLayout';
 
-import ApplicationForm from './ApplicationForm';
-import ManageApplications from './ManageApplications';
-import Dashboard from './Dashboard';
+const ApplicationForm = dynamic(() => import('./ApplicationForm'));
+const ManageApplications = dynamic(() => import('./ManageApplications'));
+const Dashboard = dynamic(() => import('./Dashboard'));
```
Comment (caravinci) [inline components/pages/Applications/index.tsx:12]: leverage the NextJS lazy loading

File: global/hooks/useApplicationsAPI.tsx [hunk 8]
```diff
@@ -17,38 +17,35 @@ const useApplicationsAPI = ({
...
   const [response, setResponse] = useState<AxiosResponse | undefined>(undefined);
-  const [isLoading, setIsLoading] = useState<boolean>(true);
   const [error, setError] = useState<AxiosError | undefined>(undefined);
 
-  const { fetchWithAuth, loadingAuth, token } = useAuthContext();
+  const { fetchWithAuth, isLoading, token } = useAuthContext();
 
   useEffect(() => {
-    token &&
+    if (token && !isLoading) {
```
Comment (caravinci) [inline global/hooks/useApplicationsAPI.tsx:29]: ensure no two concurrent calls happen from the same caller

File: global/hooks/useApplicationsAPI.tsx [hunk 9]
```diff
@@ -17,38 +17,35 @@ const useApplicationsAPI = ({
...
+          setResponse(res);
+        })
+        .catch((err: any) => {
+          setError(err);
+        });
+    }
   }, [page, pageSize, stringifySort(sort)]);
 
-  return { error, isLoading: loadingAuth || isLoading, response };
+  return { error, isLoading, response };
```
Comment (caravinci) [inline global/hooks/useApplicationsAPI.tsx:48]: simpler is better

File: global/hooks/useAuthContext.tsx [hunk 10]
```diff
@@ -70,53 +71,67 @@ export const AuthProvider = ({
...
   }: AxiosRequestConfig) => {
+    setLoading(true);
     if (!url) {
-      console.warn('no URL provided to fetchWithAuth');
-      return Promise.resolve(undefined);
+      setLoading(false);
+      return Promise.reject(undefined);
+    }
+
+    if (!token) {
```
Comment (caravinci) [inline global/hooks/useAuthContext.tsx:88]: make sure no "auth" calls happen without a token

File: global/hooks/useAuthContext.tsx [hunk 11]
```diff
@@ -70,53 +71,67 @@ export const AuthProvider = ({
   axios.defaults.headers.post['Content-Type'] = 'application/json;charset=utf-8';
   axios.defaults.headers.post['Access-Control-Allow-Origin'] = '*';
 
+  const cancelTokenSource = axios.CancelToken.source();
+  const cancelFetchWithAuth = cancelTokenSource.cancel;
```
Comment (caravinci) [inline global/hooks/useAuthContext.tsx:75]: creates request cancel functionality (not fully implemented yet)

File: global/hooks/useAuthContext.tsx [hunk 12]
```diff
@@ -70,53 +71,67 @@ export const AuthProvider = ({
...
+    token,
+    user,
   };
 
   return <AuthContext.Provider value={authData}>{children}</AuthContext.Provider>;
 };
 
 export default function useAuthContext() {
-  return React.useContext(AuthContext);
+  return useContext(AuthContext);
```
Comment (caravinci) [inline global/hooks/useAuthContext.tsx:136]: none of that `React.` nonsense lol

File: components/pages/Applications/ApplicationForm/Forms/Collaborators.tsx [hunk 13]
```diff
@@ -81,19 +81,6 @@ const mock = [
...
-  useEffect(
-    () =>
-      fetchWithAuth({
-        url: `${API.APPLICATIONS}/${appId}`,
-      })
-        .then((d) => console.log(d))
-        .catch((e) => console.error(e)),
-    [],
-  ); */
-
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/Collaborators.tsx:96]: Bonus: this belongs in the global scope, so it won't be needed here.

**Review by caravinci — COMMENTED** [review]
Help?

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**